docs/proposals/vault-auth-provider.rst

```diff
@@ -0,0 +1,154 @@
+.. vim:set ft=rst spell:
+
+Custom Vault Auth Provider for EC2 Instances - A Proposal
+===========
+
+This proposal suggests writing a custom authentication provider for Vault.
+This provider would allow access privileges, especially to the PKI secrets
+provider, to be locked down per EC2 Instance, which is not currently possible.
+
+Background
+==========
+
+Glossary
+--------
+
+-  IAM Role (iam-role) - an IAM principal. Our instances have an IAM
+   role attached which they use as a service account.
+   https://docs.aws.amazon.com/IAM/latest/UserGuide/id\_roles.html
+-  IAM Policy (iam-policy) - the authz rules (and rules documents) in
+   AWS.
+-  Vault Token (vault-token) - common identity document given by all
+   vault auth providers. Has a list of policy (documents).
+-  Vault Policy (vault-policy) - both the policy language itself (giving
+   authorization for vault operations), document of those policy
+   statement,s and the list of such documents named on a vault-token.
+   vault-policy has its own type, which we'll call *policy*
+-  Vault PKI Role (pki-role) - vault-policy isn't very powerful wrt to
+   PKI operations, e.g. it doesn't let you lock PKI certificate issuance
+   down to specific CNs etc. The PKI secrets provider has its own
+   mechanism for this; the PKI role. pki-roles live at paths and access
+   is controlled by vault-policy. pki-roles are implemented as vault
+   generic secrets.
+   https://www.vaultproject.io/api/secret/pki/index.html#create-update-role
+
+Problem
+-------
+
+We would like to have individual Vault policies per EC2 instance, so
+that they can't issue themselves certificates for anyone else's FQDN.
+
+See:
+
+- https://github.com/jetstack/tarmak/issues/120
+- https://github.com/jetstack/tarmak/issues/34
+
+Posibilities
+------------
+
+Due to several Vault limitations, especially in the free version, it's
+not possible to:
+
+- Add a vault-policy (name) to a token after it's been created (by vault ec2 login)
+- Specify a vault-policy expressive enough to re-use one policy for all EC2 Instances (i.e. no variable back-references)
+- Vault Premium has a different policy language called Sentinel, which looks like it could do this
```

Comment:
Depending on the difficulty/reusability of the existing AWS auth provider, could we opt for a simpler implementation that requires Vault Enterprise to use?

I'm not sure whether this can be considered a feature that would be largely of interest to enterprise, and whilst it's nice to have supporting the free version the onus of maintaining an auth plugin for general public consumption should be considered.

Reply:
Vault enterprise's sentinel seems to only have access to the meta data available in the token's metadata. The metadata doesn't contain the IP address or the hostname, so I think even with Vault enterprise we would need to do something to get IP/hostname into the token's metadata

Reply:
We'd then also have to get everyone on Vault Enterprise, as I *really* wouldn't wanna support two versions